### ISSUE TITLE
Pattern validation fails when START STOP timestamps have a leading 't'

### ISSUE THREAD

**Issue by delliott90** [oasis-open/cti-pattern-validator#52]
Using the qualifier example from the STIX 2 documentation...

`Enter a pattern to validate: [file:name = 'foo.dll'] START t'2016-06-01T00:00:00Z' STOP t'2016-07-01T00:00:00Z'`
`FAIL: Error found at line 1:30. mismatched input 't'2016-06-01T00:00:00Z'' expecting StringLiteral`

**Comment by clenk** [comment]
Thanks @delliott90; looks like this is a bug in our pattern grammar that we need to fix.

**Comment by chisholm** [comment]
I think it's a bug in the documentation.  The STIX 2.0 (part 5) spec says:

```
... MUST be a timestamp as defined in section 2.10 of STIX™ Version 2.0. Part 1: STIX Core Concepts.
```

The part 1 spec says about the timestamp type:

```
The timestamp property MUST be a valid RFC 3339-formatted timestamp [RFC3339]
```

and gives the example:

```
{
  ...
  "created": "2016-01-20T12:31:12.123Z",
  ...
}
```

Note that there is no "t" prefix in the example (and can't be if they're delegating to part 1, since that's a patterning-specific syntax).  In fact, I remember thinking about this when I was writing the [matcher](https://github.com/oasis-open/cti-pattern-matcher/blob/0e1d6e43a3af433748c4d7bfe54d6a0574c31440/stix2matcher/matcher.py#L1355).  I thought it might have made sense to use timestamp literals there, but they didn't.  That was my understanding of it anyway.

**Comment by chisholm** [comment]
Ohhh... they changed it in the STIX **2.1** spec!  (You just said STIX 2, so I was thinking 2.0.)  Ok, add another thing to the list when we get around to writing a 2.1-compliant grammar, validator, matcher, etc. :)  (I don't think any of those pattern-related tools have had a 2.1 update yet, have they?)

**Comment by clenk** [comment]
Yes, @chisholm's right. These libraries aren't 2.1 compliant yet.

**Comment by JasonKeirstead** [comment]
Hi @clenk @chisholm , having the 't' in front of timestamp is actually part of 2.0. 

See the CS: 

A constant of timestamp type encodes a STIX timestamp (as specified in section 2.10 of STIX™ Version 2.0 Part 1: STIX Core Concepts) as a string. The timestamp string MUST be surrounded by apostrophes (''' U+0027) and prefixed with a 't' (U+0074).

Example:

t'2014-01-13T07:03:17Z'

**Comment by JasonKeirstead** [comment]
The bug is in the grammar. START and STOP should be followed by TimestampLiteral, not StringLiteral.  TimestampLiteral is defined in the grammar and actually never used, currently. It is meant to be used there.

**Comment by chisholm** [comment]
So just for clarity and posterity, what I understand the problem to be is that the STIX 2.0 part 5 spec has a mistake.  You quoted the section 2.1 documentation for the timestamp constant, but the section 4.1.1 documentation of the START/STOP qualifier doesn't refer to the timestamp constant.  It specifically refers to the part 1 timestamp type.  It should refer to the part 5 constant, and that has been fixed in STIX 2.1.

So, the grammar follows the letter of the spec, but not the intent (which I hadn't realized until now).  The question then becomes, which should we follow?  Sounds like you advocate for following the intent.

**Comment by JasonKeirstead** [comment]
@chisholm No I am quoting the 2.0 Part 5 CSD. It seems like in the 2.0 spec there are two, contradictory definitions of how to enter timestamps... one in START/STOP and another in 'Constants'

OK I agree its a 2.1 issue.

**Comment by chisholm** [comment]
Ergh... I meant to say that you quoted STIX 2.0.  That was my typo, sorry!  I have edited my post accordingly.  So much for clarity... \*sigh\*.

**Comment by delliott90** [comment]
@clenk I can make a fix to the grammar file. I notice that the schemas project has a master and stix2.1 branch. Which one should I make the fix against?

**Comment by clenk** [comment]
@delliott90 that would be great! Make it against the stix2.1 branch, please.
